File: engine/leaderboard_engine.py
"""
GridForge Leaderboard Engine — Python port of src/engine/LeaderboardEngine.js

Builds multi-dimensional leaderboard from player scores.
"""

from __future__ import annotations
import math


def build_leaderboard(players: list[dict]) -> dict:
    if not players:
        return {"overall": [], "roleWinners": {}, "systemSteward": None, "mostConsistent": None}

    overall = sorted(
        players,
        key=lambda p: (-(p.get("overallScore", 0)), -(p.get("roleScore", 0)), -(p.get("cash", 0))),
    )
    overall = [{**p, "rank": i + 1} for i, p in enumerate(overall)]

    # Role winners
    by_role: dict[str, list[dict]] = {}
    for p in players:
        r = p.get("role", "GENERATOR")
        by_role.setdefault(r, []).append(p)

    role_winners = {}
    for role, role_players in by_role.items():
        best = max(role_players, key=lambda x: x.get("roleScore", 0))
        if best.get("roleScore", 0) > 0:
            role_winners[role] = {"id": best["id"], "name": best["name"], "roleScore": best["roleScore"]}

    # System steward
    system_steward = max(players, key=lambda p: p.get("systemScore", 0))

    # Most consistent
    most_consistent = None
    if players and players[0].get("scoreHistory") and len(players[0]["scoreHistory"]) > 1:
        with_variance = []
        for p in players:
            scores = p.get("scoreHistory", [])
            mean = sum(scores) / len(scores) if scores else 0
            variance = sum((s - mean) ** 2 for s in scores) / len(scores) if scores else 0
            with_variance.append({**p, "variance": variance})
        most_consistent = min(with_variance, key=lambda x: x["variance"])
    else:
        avg_score = sum(p.get("overallScore", 0) for p in players) / len(players)
        most_consistent = min(players, key=lambda p: abs(p.get("overallScore", 0) - avg_score))

    return {"overall": overall, "roleWinners": role_winners, "systemSteward": system_steward, "mostConsistent": most_consistent}


def get_rank_label(rank: int) -> str:
    if rank == 1:
        return "🥇"
    if rank == 2:
        return "🥈"
    if rank == 3:
        return "🥉"
    return f"#{rank}"


def get_score_color(score: int) -> str:
    if score >= 80:
        return "#1de98b"
    if score >= 60:
        return "#38c0fc"
    if score >= 40:
        return "#f5b222"
    if score >= 20:
        return "#f0855a"
    return "#f0455a"


def generate_player_narrative(player: dict | None) -> str:
    if not player:
        return ""

    rs = player.get("roleScore", 0)
    ss = player.get("systemScore", 0)
    role_detail = player.get("roleDetail", {})
    primary_name = role_detail.get("primary", {}).get("name", "Performance") if role_detail else "Performance"
    primary_val = role_detail.get("primary", {}).get("value", "—") if role_detail else "—"

    if rs >= 85:
        tier = "exceptional"
    elif rs >= 70:
        tier = "strong"
    elif rs >= 50:
        tier = "decent"
    elif rs >= 30:
        tier = "struggling"
    else:
        tier = "poor"

    system_note = ""
    if ss >= 80:
        system_note = " — excellent system citizen"
    elif ss >= 60:
        system_note = " — helpful to the grid"
    elif ss <= 30:
        system_note = " — often destabilising"

    role_name = player.get("role", "Player")
    cash = round(player.get("cash", 0))
    return f"{role_name}: {tier} {primary_name} ({primary_val}){system_note}. P&L: £{cash}."


def build_round_debrief(leaderboard_data: dict, system_state: dict) -> dict:
    overall = leaderboard_data.get("overall", [])
    role_winners = leaderboard_data.get("roleWinners", {})
    system_steward = leaderboard_data.get("systemSteward")

    niv_history = system_state.get("nivHistory", [])
    avg_abs_niv = (
        round(sum(e.get("absNiv", 0) for e in niv_history) / len(niv_history))
        if niv_history
        else 0
    )

    return {
        "podium": overall[:3],
        "roleWinners": role_winners,
        "systemSteward": {
            "id": system_steward["id"],
            "name": system_steward["name"],
            "systemScore": system_steward.get("systemScore"),
        } if system_steward else None,
        "systemMetrics": {
            "avgAbsNIV": avg_abs_niv,
            "totalBalancingCost": round(system_state.get("totalBalancingCost", 0)),
            "stressEvents": system_state.get("stressEvents", 0),
            "blackouts": system_state.get("blackouts", 0),
            "totalSPs": system_state.get("totalSPs", 0),
        },
        "narratives": [
            {"id": p["id"], "name": p["name"], "narrative": generate_player_narrative(p)}
            for p in overall[:5]
        ],
    }

File: frontend/src/engine/LeaderboardEngine.js
// ─── LEADERBOARD ENGINE ───
// Builds multi-dimensional leaderboard from player scores.
// Supports: overall ranking, role winners, system steward, most consistent.

// ─── Build Full Leaderboard ───
// players: array of { id, name, role, roleScore, systemScore, overallScore, roleDetail, cash }
// Returns: { overall: [...sorted], roleWinners: {}, systemSteward, mostConsistent }
export function buildLeaderboard(players) {
    if (!players || players.length === 0) {
        return { overall: [], roleWinners: {}, systemSteward: null, mostConsistent: null };
    }

    // Overall ranking: sort by overallScore descending, tie-break by roleScore, then cash
    const overall = [...players].sort((a, b) => {
        if ((b.overallScore || 0) !== (a.overallScore || 0)) return (b.overallScore || 0) - (a.overallScore || 0);
        if ((b.roleScore || 0) !== (a.roleScore || 0)) return (b.roleScore || 0) - (a.roleScore || 0);
        return (b.cash || 0) - (a.cash || 0);
    }).map((p, i) => ({ ...p, rank: i + 1 }));

    // Role winners: best roleScore within each role
    const roleWinners = {};
    const byRole = {};
    for (const p of players) {
        const r = p.role || 'GENERATOR';
        if (!byRole[r]) byRole[r] = [];
        byRole[r].push(p);
    }
    for (const [role, rolePlayers] of Object.entries(byRole)) {
        const best = rolePlayers.reduce((a, b) => (b.roleScore || 0) > (a.roleScore || 0) ? b : a, rolePlayers[0]);
        if (best && (best.roleScore || 0) > 0) {
            roleWinners[role] = { id: best.id, name: best.name, roleScore: best.roleScore };
        }
    }

    // System steward: highest systemScore — only award if at least one player has a positive score
    const bestSysScore = Math.max(...players.map(p => p.systemScore || 0));
    const systemSteward = bestSysScore > 0
        ? players.reduce((a, b) => (b.systemScore || 0) > (a.systemScore || 0) ? b : a, players[0])
        : null;

    // Most consistent: lowest variance in overallScore history (if available)
    // If no history, use the player with overallScore closest to the group average
    let mostConsistent;
    if (players[0] && players[0].scoreHistory && players[0].scoreHistory.length > 1) {
        // Calculate variance for each player
        const withVariance = players.map(p => {
            const scores = p.scoreHistory;
            const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
            const variance = scores.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / scores.length;
            return { ...p, variance };
        });
        mostConsistent = withVariance.reduce((a, b) => a.variance < b.variance ? a : b);
    } else {
        // Fallback: player with overallScore closest to group average
        const avgScore = players.reduce((sum, p) => sum + (p.overallScore || 0), 0) / players.length;
        mostConsistent = players.reduce((a, b) => {
            const aDiff = Math.abs((a.overallScore || 0) - avgScore);
            const bDiff = Math.abs((b.overallScore || 0) - avgScore);
            return aDiff < bDiff ? a : b;
        });
    }

    return { overall, roleWinners, systemSteward, mostConsistent };
}

// ─── Rank Label ───
export function getRankLabel(rank) {
    if (rank === 1) return '🥇';
    if (rank === 2) return '🥈';
    if (rank === 3) return '🥉';
    return `#${rank}`;
}

// ─── Score Color ───
export function getScoreColor(score) {
    if (score >= 80) return '#1de98b';
    if (score >= 60) return '#38c0fc';
    if (score >= 40) return '#f5b222';
    if (score >= 20) return '#f0855a';
    return '#f0455a';
}

// ─── Generate Player Narrative ───
// One-sentence summary of what drove this player's score
export function generatePlayerNarrative(player) {
    if (!player) return '';

    const { role, roleScore, systemScore, overallScore, roleDetail, cash } = player;
    const rs = roleScore || 0;
    const ss = systemScore || 0;

    // Primary KPI description
    const primaryName = roleDetail?.primary?.name || 'Performance';
    const primaryVal = roleDetail?.primary?.value ?? '—';

    // Determine performance tier
    let tier = 'solid';
    if (rs >= 85) tier = 'exceptional';
    else if (rs >= 70) tier = 'strong';
    else if (rs >= 50) tier = 'decent';
    else if (rs >= 30) tier = 'struggling';
    else tier = 'poor';

    // System behavior
    let systemNote = '';
    if (ss >= 80) systemNote = ' — excellent system citizen';
    else if (ss >= 60) systemNote = ' — helpful to the grid';
    else if (ss <= 30) systemNote = ' — often destabilising';

    const roleName = role || 'Player';
    return `${roleName}: ${tier} ${primaryName} (${primaryVal})${systemNote}. P&L: £${Math.round(cash || 0)}.`;
}

// ─── Build Round Debrief Data ───
// Returns structured data for the round debrief overlay
export function buildRoundDebrief(leaderboardData, systemState) {
    const { overall, roleWinners, systemSteward } = leaderboardData;

    return {
        podium: overall.slice(0, 3),
        roleWinners,
        systemSteward: systemSteward ? {
            id: systemSteward.id,
            name: systemSteward.name,
            systemScore: systemSteward.systemScore,
        } : null,
        systemMetrics: {
            avgAbsNIV: systemState.nivHistory.length > 0
                ? (systemState.nivHistory.reduce((s, e) => s + e.absNiv, 0) / systemState.nivHistory.length).toFixed(0)
                : 0,
            totalBalancingCost: Math.round(systemState.totalBalancingCost || 0),
            stressEvents: systemState.stressEvents || 0,
            blackouts: systemState.blackouts || 0,
            totalSPs: systemState.totalSPs || 0,
        },
        narratives: overall.slice(0, 5).map(p => ({
            id: p.id,
            name: p.name,
            narrative: generatePlayerNarrative(p),
        })),
    };
}
